### PR TITLE
Introduce top-level schema configuration directives

### PR DESCRIPTION
**Pull request by MacondoExpress**
# Description

Introduced the top-level schema configuration directives, and introduce schema configuration helper functions to help to configure the output of "makeAugmentedSchema".
Also, includes some fixes for a  couple of typos.

## Complexity

Complexity: Low

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [x] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-3333.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 46 | 54 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 82 | 109 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 60 | 63 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 4219 | 3886 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 4189 | 3423 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 52 | 80 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 55 | 70 | 2174 |
|  aggregations.DeeplyNestedCount | 13070329 | 13070331 | 5808 | 6082 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 177 | 274 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 89 | 78 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 309 | 295 | 3003 |
|  connections.Connection | 14082 | 14082 | 95 | 84 | 2174 |
|  connections.NestedConnection | 45470 | 45472 | 153 | 159 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 96 | 106 | 1040 |
|  connections.ConnectionWithSortAndCypher | 3284 | 3284 | 340 | 514 | 1040 |
|  create.SimpleMutation | 7 | 7 | 72 | 110 | 1 |
|  cypher-directive.TopLevelCypherDirective | 122100 | 122100 | 716 | 849 | 12241 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 153617 | 153617 | 218 | 237 | 12241 |
|  delete.SimpleDelete | 19401 | 19401 | 331 | 430 | 1040 |
|  delete.NestedDeleteInUpdate | 23202 | 23202 | 327 | 220 | 2040 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 91 | 53 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 125 | 145 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 3450 | 3450 | 41 | 41 | 1134 |
|  2925.NestedSingleRelationshipRequiredFilter | 9383 | 9383 | 80 | 79 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 26 | 32 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 58 | 72 | 2174 |
|  query.QueryWhere | 9707 | 9722 | 44 | 57 | 2164 |
|  query.SimpleQueryWithNestedWhere | 9885 | 9900 | 64 | 66 | 2164 |
|  query.Nested | 10096041 | 10096041 | 12258 | 11503 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 12179 | 11703 | 2004000 |
|  query.OrFilterOnRelationships | 43202 | 42795 | 234 | 293 | 2020 |
|  query.OrFilterOnRelationshipsAndNested | 36855 | 36915 | 323 | 349 | 2020 |
|  query.QueryWithNestedIn | 13546 | 13656 | 83 | 72 | 1263 |
|  query.NestedConnectionWhere | 9834 | 9834 | 68 | 75 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10069 | 9906 | 99 | 168 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19302 | 19086 | 195 | 252 | 1652 |
|  query.NestedWithRelationshipSingleFilters | 3732 | 3732 | 405 | 418 | 1134 |
|  query.Fulltext | 80 | 80 | 92 | 62 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 63 | 71 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 87 | 84 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1482 | 1414 | 327 | 434 | 1040 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 90 | 77 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 136 | 122 | 4516 |
|  sorting.SortWithTopLevelCypher | 3121 | 3121 | 247 | 343 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 85 | 79 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 83 | 77 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 370 | 387 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 342 | 489 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 145 | 161 | 2002 |

</details>
        
Old Schema Generation: 55.534s
Schema Generation: 55.994s
Old Subgraph Schema Generation: 1:15.872 (m:ss.mmm)
Subgraph Schema Generation: 1:07.380 (m:ss.mmm)

File: packages/graphql/src/schema/schema-configuration.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            if (directive.name.value === mutationDirectiveDefinition.name) {
+                if (schemaConfiguration.mutationDirective) {
+                    throw new Error(`Ambiguity usage with the directive named: ${queryDirectiveDefinition.name}`);
+                }
+                schemaConfiguration.mutationDirective = parseMutationDirective(directive);
+            }
+
+            if (directive.name.value === subscriptionDirectiveDefinition.name) {
+                if (schemaConfiguration.subscriptionDirective) {
+                    throw new Error(`Ambiguity usage with the directive named: ${queryDirectiveDefinition.name}`);
```
Comment (mjfwebb) [inline packages/graphql/src/schema/schema-configuration.ts:90]:
```suggestion
                    throw new Error(`Ambiguous usage with the directive named: ${queryDirectiveDefinition.name}`);
```

File: packages/graphql/tests/schema/directives/mutation.test.ts [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+            const updateActors = mutationFields["updateActors"];
+
+            expect(updateMovies).toBeDefined();
+            expect(updateActors).toBeDefined();
+
+            const deleteMovies = mutationFields["deleteMovies"];
+            const deleteActors = mutationFields["deleteActors"];
+
+            expect(deleteMovies).toBeDefined();
+            expect(deleteActors).toBeDefined();
```
[the remark this answers is not in the document]
Comment (MacondoExpress) [reply packages/graphql/tests/schema/directives/mutation.test.ts:58]: I'm not sure about this, in this case, all fields are defined but in this test file as well as in other schema configuration tests files is common to have some fields defined and others not. That means that adopting this strategy will likely reach something among these lines:

```ts
const definedFieldNames = [
    "createMovies",
    "updateMovies",
    "deleteMovies",
    "deleteActors",
];

for (const fieldName of definedFieldNames) {
    expect(mutationFields[fieldName]).toBeDefined();
}
const undefinedFieldNames = [
    "createActors",
    "updateActors"
];

for (const fieldName of undefinedFieldNames) {
    expect(mutationFields[fieldName]).toBeDefined();
}
```
Also, in my opinion, if the assertion 
```ts
expect(mutationFields[fieldName]).toBeDefined();
``` 
fails it provides less information immediately compared to for instance: 
```ts
expect(createMovies).toBeUndefined();
```

[the remark this answers is not in the document]
Comment (mjfwebb) [reply packages/graphql/tests/schema/directives/mutation.test.ts:58]: Yeah, you make good points!

File: packages/graphql/src/classes/MutationDirective.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+export class MutationDirective {
+    public create: boolean;
+    public update: boolean;
+    public delete: boolean;
+
+    constructor(operations?: ("CREATE" | "UPDATE" | "DELETE")[]) {
+        const operationsSet = new Set(operations);
+        this.create = operationsSet.has("CREATE");
+        this.update = operationsSet.has("UPDATE");
+        this.delete = operationsSet.has("DELETE");
```
[the remark this answers is not in the document]
Comment (MacondoExpress) [reply packages/graphql/src/classes/MutationDirective.ts:29]: ty!

File: packages/graphql/src/utils/getArgumentValues.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ */
+export function getArgumentValues(
+    def: GraphQLField<unknown, unknown> | GraphQLDirective,
+    node: FieldNode | DirectiveNode,
+    variableValues?: Maybe<ObjMap<unknown>>
+): { [argument: string]: unknown } {
+    const coercedValues: { [argument: string]: unknown } = {};
+
+    // FIXME: https://github.com/graphql/graphql-js/issues/2203
+    /* c8 ignore next */
```
[the remark this answers is not in the document]
Comment (MacondoExpress) [reply packages/graphql/src/utils/getArgumentValues.ts]: Not related to us, is an active bug from the GraphQL-JS repository, we will start using the exported method when we will drop support to v15

**Review by mjfwebb — APPROVED** [review]
Good stuff :+1:

**Pull request merged**